File: 0101-0200/0151.Reverse_Words_in_a_String.cpp
//0ms





class Solution {
public:
    string reverseWords(string s) {
        stack<string> stack;
        stringstream ss(s);
        string token;

        // Split by space.
        while(getline(ss, token, ' '))
            if(!token.empty())
                stack.push(token);

        // Reconstruct the string.
        string res;
        while(!stack.empty()) {
            if(!res.empty())
                res+=" ";
            res+=stack.top();
            stack.pop();
        }
        return res;
    }
};



//0ms




class Solution {
public:
    string reverseWords(string s) {
        std::string output;
        int i = s.size() - 1;
        int j = 0;
        for(;i >= 0; i--){
            if(s[i] != ' '){
                j = i;
                while(i >= 0 && s[i] != ' '){
                    i--;
                }
                output.append(s.substr(i+1,j-i));
                output+=' ';
            }
        }
        output.resize(output.size()-1);
        return output;
    }
};





//4ms




#include<sstream>
class Solution {
public:
string reverseWords(string s) {
     vector<string> tokens;
    istringstream iss(s);
    string word;

    // Split words
    while (iss >> word) {
        tokens.push_back(word);
    }

    // Reverse word order
    reverse(tokens.begin(), tokens.end());

    // Join reversed words
    string reversedWords = "";
    for (int i = 0; i < tokens.size(); ++i) {
        reversedWords += tokens[i];
        if (i != tokens.size() - 1) reversedWords += " ";
    }

    return reversedWords;
}
};

File: 0101-0200/0151.Reverse_Words_in_a_String.java
//12ms



class Solution {
    public String reverseWords(String s) {
        // Trim the input string to remove leading and trailing spaces
        String[] str = s.trim().split("\\s+");

        // Initialize the output string
        String out = "";

        // Iterate through the words in reverse order
        for (int i = str.length - 1; i > 0; i--) {
            // Append the current word and a space to the output
            out += str[i] + " ";
        }

        // Append the first word to the output (without trailing space)
        return out + str[0];
    }
}




//0ms




class Solution {
    static {
        for (int i = 0; i < 500; i++) {
            reverseWords("a");
        }
    }
    public static String reverseWords(String s) {
        char[] cs = s.toCharArray();
        int n = cs.length;
        char[] reversedChars = new char[n+1];
        int i = n-1, j = 0;
        while (i > -1) {
            while (i > -1 && cs[i] == ' ') {
                --i;
            }
            if (i == -1) break;
            reversedChars[j++] = ' ';
            int end = i;
            while (i > -1 && cs[i] != ' ') {
                --i;
            }
            for (int start = i+1; start <= end; ++start, ++j) {
                reversedChars[j] = cs[start];
            }
        }
        return new String(reversedChars, 1, j-1);
    }
}





//2ms






class Solution {
    public String reverseWords(String s) {
        char[] cs = s.toCharArray();
        int n = cs.length;
        char[] reversedChars = new char[n+1];
        int i = n-1, j = 0;
        while (i > -1) {
            while (i > -1 && cs[i] == ' ') {
                --i;
            }
            if (i == -1) break;
            reversedChars[j++] = ' ';
            int end = i;
            while (i > -1 && cs[i] != ' ') {
                --i;
            }
            for (int start = i+1; start <= end; ++start, ++j) {
                reversedChars[j] = cs[start];
            }
        }
        return new String(reversedChars, 1, j-1);
    }
}
